sc-memory/sc-memory/src/scs/scs_parser.cpp

```diff
@@ -413,44 +456,94 @@ ElementHandle Parser::ProcessIdentifier(std::string const & name)
-  elType |= scs::TypeResolver::IsConst(name) ? ScType::Const : ScType::Var;
-
-  return AppendElement(name, elType);
+  ScType type = scs::TypeResolver::GetKeynodeType(scsType);
+  type |= scs::TypeResolver::IsConst(name) ? ScType::Const : ScType::Var;
+  return AppendElement(name, type);
 }
 
-void Parser::ProcessTriple(ElementHandle const & source, ElementHandle const & connector, ElementHandle const & target)
+void Parser::ProcessTriple(
+    ElementHandle const & sourceHandle,
+    ElementHandle const & connectorHandle,
+    ElementHandle const & targetHandle)
 {
-  ParsedElement const & connectorEl = GetParsedElement(connector);
+  ParsedElement & connector = GetParsedElementRef(connectorHandle);
+
+  auto AddConnector =
+      [this, &connector](
+          ElementHandle const & sourceHdl, ElementHandle const & connectorHdl, ElementHandle const & targetHdl)
+  {
+    ParsedElement const & source = GetParsedElement(sourceHdl);
+    ParsedElement & target = GetParsedElementRef(targetHdl);
 
-  auto AddConnector = [this, &connectorEl](ElementHandle const & src, ElementHandle const & e, ElementHandle const & trg) {
-    ParsedElement const & srcEl = GetParsedElement(src);
-    std::string const & idtf = srcEl.GetIdtf();
-    if (connectorEl.GetType() == ScType::ConstPermPosArc && scs::TypeResolver::IsKeynodeType(idtf))
+    if (source.IsElementType() && (connector.GetType() == ScType::ConstPermPosArc))
     {
-      ParsedElement & targetEl = GetParsedElementRef(trg);
-      ScType const newType = targetEl.m_type | scs::TypeResolver::GetKeynodeType(idtf);
+      std::string const & sourceIdtf = source.GetIdtf();
+      ScType const newType = target.m_type | scs::TypeResolver::GetKeynodeType(sourceIdtf);
 
-      if (targetEl.m_type.CanExtendTo(newType))
-        targetEl.m_type = newType;
+      if (target.m_type.CanExtendTo(newType))
+        target.m_type = newType;
       else
-        SC_THROW_EXCEPTION(utils::ExceptionParseError, "Can't merge types for element " + targetEl.GetIdtf());
-
-      if (!m_contourTriplesStack.empty())
-        m_parsedTriples.emplace_back(src, e, trg);
+        SC_THROW_EXCEPTION(
+            utils::ExceptionParseError,
+            "Can't extend type `" << std::string(target.m_type) << "` to type `" << std::string(newType)
+                                  << "` for element `" << target.GetIdtf() << "`.");
+
+      // TODO(NikitaZotov): Unfortunately, parser collects all sc.s-elements, and only then forms sc.s-triples based on
+      // the parsed sc.s-elements. Due to this, it is difficult to handle cases when it is necessary not to generate a
+      // sc-arc from a type of sc-elements to sc-element. Therefore, now the parser memorizes these arcs after they were
+      // parsed.
+      m_elementTypeOutgoingBaseArcs.insert(connector.GetIdtf());
+    }
+    else if (source.IsElementType() || target.IsElementType())
+    {
+      ElementHandle const & handle = source.IsElementType() ? sourceHdl : targetHdl;
```

Comment:
I wanted to check what happens when both source and target are element types and it is something unpredicted(tuple belongs to structure 2 times and class does not belong to structure)
![image](https://github.com/user-attachments/assets/d16e0cd7-3971-4c75-8551-fde432451ff4)

![image](https://github.com/user-attachments/assets/e92a2687-cc16-48b9-b280-a723720af807)